modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvPartitionStorageConcurrencyTest.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.storage;
+
+import static org.apache.ignite.internal.testframework.IgniteTestUtils.startRace;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
+
+import org.apache.ignite.internal.hlc.HybridTimestamp;
+import org.apache.ignite.internal.storage.impl.TestStorageEngine;
+import org.junit.jupiter.api.RepeatedTest;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Test to check for race conditions in MV partition storage.
+ */
+public abstract class AbstractMvPartitionStorageConcurrencyTest extends BaseMvPartitionStorageTest {
+    /** To be used in a loop. {@link RepeatedTest} has a smaller failure rate for some reasons. */
+    private static final int REPEATS = 100;
+
+    @Test
+    void testAbortAndRead() throws Exception {
+        for (int i = 0; i < REPEATS; i++) {
+            addWrite(ROW_ID, BINARY_ROW, TX_ID);
+
+            startRace(
+                    () -> abortWrite(ROW_ID),
+                    () -> read(ROW_ID, clock.now()),
+                    () -> scanFirstEntry(clock.now())
+            );
+
+            assertNull(read(ROW_ID, clock.now()));
+        }
+    }
+
+    @Test
+    void testCommitAndRead() throws Exception {
+        for (int i = 0; i < REPEATS; i++) {
+            addWrite(ROW_ID, BINARY_ROW, TX_ID);
+
+            startRace(
+                    () -> commitWrite(ROW_ID, clock.now()),
+                    () -> read(ROW_ID, clock.now()),
+                    () -> scanFirstEntry(clock.now())
+            );
+
+            assertRowMatches(read(ROW_ID, clock.now()), BINARY_ROW);
+        }
+    }
+
+    @Test
+    void testUpdateAndRead() throws Exception {
+        for (int i = 0; i < REPEATS; i++) {
+            addWrite(ROW_ID, BINARY_ROW, TX_ID);
+
+            startRace(
+                    () -> addWrite(ROW_ID, BINARY_ROW, TX_ID),
```

Comment:
That would be a different scenario, right? Here I specifically want to update existing write intent.

Reply:
My concern is about a different row (different value), but with same rowId. This would update the existing write intent, but it will change its value. In the current test the value after change is same, so it's impossible to say whether the update was applied or not.

Reply:
Ah, ok, I see